text/trace/0235-sampling-threshold-in-trace-state.md

```diff
@@ -0,0 +1,87 @@
+This proposal allows for the continued expression of randomness using `r-value` as specified there using the key `r`.
+However, that value is limited to powers of two, while this proposal is not; to distinguish the cases, this proposal uses the key `rv`.
+
+In order to make consistent sampling decisions across the entire path of the trace, two values SHOULD be propagated with the trace:
+
+1. A random (or pseudo-random) value of at least 56 bits, called `R` below.
+2. A 56-bit trace threshold as expressed in the TraceState, called `T` below.
+
+The sampling decision is propagated with the following algorithm:
+
+* If the `th` key is not specified, Always Sample.
+* Else derive `T` by parsing the `th` key as a hex value as described below.
+* If `T` is 0, Always Sample. This implies that non-probabalistic sampling is taking place.
+* Compare the 56 bits of `T` with the 56 bits of `R`. If `T <= R`, then do not sample.
+
+The `R` value SHALL be derived as follows:
+
+* If the key `rv` is present in the Tracestate header, then `R = rv`.
+* Else if the Random Trace ID Flag is `true` in the traceparent header, then `R` is the lowest-order 56 bits of the trace-id.
+* Else `R` SHALL be generated as a random value in the range `(0, (2**56)-1)` and added to the Tracestate header with key `rv`.
+
+The preferred way to propagate the `R` value is as the lowest 56 bits of the trace-id.
+If these bits are in fact random, the `random` trace-flag SHOULD be set as specified in [the W3C trace context specification](https://w3c.github.io/trace-context/#trace-id).
+There are circumstances where trace-id randomness is inadequate (for example, sampling a group of traces together); in these cases, an `rv` value is required.
+
+The value of the `rv` and `th` keys SHALL be expressed as up to 14 hexadecimal characters from the set `[0-9a-f]`. Trailing zeros (but not leading zeros) may be omitted. _Question: should we permit [A-F] as well, or instead?_
+
+Examples:
+`th` value is missing: Always Sample (probability = 100%). The AlwaysSample sampler in the OTel SDK should do this.
+`th=8` -- equivalent to `th=80000000000000`, which is 50% probability.
+`th=08` -- equivalent to `th=08000000000000`, which is 3.125% probability.
+`th=0` -- equivalent to `th=00000000000000`, which means Always Sample; this is outside of probabalistic sampling.
+
+The `T` value SHALL be derived as follows:
+
+* If the `th` key is not present in the Tracestate header, then `T` is effectively 2^56 (which doesn't fit in 56 bits).
+* Else the value corresponding to the `th` key should be interpreted as above.
+
+Sampling Decisions SHOULD be propagated by setting the value of the `th` key in the Tracestate header according to the above.
+
+In the case of a downstream sampler that is attempting to reduce the volume of traffic, the sampler MAY modify the `th` header.
+Note that the `T` value may be reduced by a downstream sampler, but should never be increased.
```

Comment:
Could you please elaborate why it cannot be increased? It seems like that would prevent a common use case of inflationary sampling.

Reply:
Some rewording is needed here -- I think by "downstream" @kentquirk means on the collection path, i.e., in a tail-sampling scenario. In tail sampling, to reduce the threshold means to increase sampling probability after the fact, which is a lie. I think this language can be as strong as `MUST NOT`, for tail sampling.

But note that it's OK for head samplers. A non-root sampler can raise or lower sampling threshold with the repercussions discussed e.g., in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#recommenendation-use-non-descending-power-of-two-probabilities. If a non-root head sampler lowers probability, you may expect incomplete traces (missing their children, difficult to detect!), and if a non-root head sampler raises probability, you. may expect complete subtraces (missing their parents, easy to detect).

Reply:
I took a shot at it. Please tell me what you think.